InitialSessionState: Add support for configuring initial working directory

MatejKafka opened the pull request:
# PR Summary

Add support for configuring the initial working location of a `Runspace` using a new ~`WorkingDirectory`~`Location` property on `InitialSessionState`.

I recommend reviewing the 2 commits separately, first one refactors the method without major changes in behavior, second one implements the feature.

## PR Context

Resolves #17603.

## Design questions

@vexx32 I have 2 questions about the design of the feature and related functionality:

1\) If the working directory is invalid (does not exist, unknown drive,...), I propagate the exception back to the caller, without checking `ThrowOnRunspaceOpenError`. Should I add a check and suppress the exception if it's not set? My reasoning is that if a caller explicitly sets a custom working directory, he probably expects it to be valid and get notified when it's not, even when he's not interested in other initialization errors, but it's inconsistent with how other errors are handled.

2\) In the original implementation, there's a check for `if (context.EngineSessionState.ProviderCount > 0)` and if there are no providers, the function does not do anything, leaving the current drive and location unset. However, this results in subsequent accesses to the location throwing confusing internal errors (even from scripts running inside the `Runspace`):
```powershell
$iss = [initialsessionstate]::CreateDefault2()
$iss.ThrowOnRunspaceOpenError = $true
# remove all providers
$iss.Providers.RemoveItem(0, $iss.Providers.Count)

$r = [runspacefactory]::CreateRunspace($Host, $iss)
$r.Open()
[powershell]::Create($r).AddCommand("pwd").Invoke()
# Exception calling "Invoke" with "0" argument(s): "Cannot perform operation because operation "get_CurrentLocation" is not valid. Remove operation "get_CurrentLocation", or investigate why it is not valid."
```

For now, I've kept the original behavior, but it seems cleaner to me to throw an exception if there are no providers defined to avoid the confusing errors later on. I can either modify the behavior in a subsequent commit, or I can create a new PR after this one is done, just wanted to hear your opinion.


## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

MatejKafka committed:
InitialSessionState: Refactor `SetSessionStateDrive` in preparation for adding working directory configuration support (see next commit)

Behavior should be unchanged, except that the catch-all `try` block was shortened to the relevant part of the function – exceptions thrown by the rest of the function are unrelated to drive and location configuration and indicate serious issues which should probably be propagated.

MatejKafka committed:
InitialSessionState: Add support for configuring initial working directory (#17603)

jborean93 commented on src/System.Management.Automation/engine/InitialSessionState.cs:
```diff
@@ -1831,6 +1832,12 @@ public Microsoft.PowerShell.ExecutionPolicy ExecutionPolicy
+        /// If not null, the working directory of the runspace is set to this path. If null,
+        /// the <see cref="Environment.CurrentDirectory">process working directory</see> is used as a default.
+        /// </summary>
+        public string WorkingDirectory { get; set; } = null;
```
As this is a PowerShell location this probably should be called `Location` or `DefaultLocation` (or something along those lines). The location can be more than just a filesystem path so the API should probably not tie itself to that naming standard.

MatejKafka replied:
Good point. I used the name I mentioned in the original issue, but `DefaultLocation` is better.

jborean93 commented on src/System.Management.Automation/engine/InitialSessionState.cs:
```diff
@@ -3381,77 +3390,87 @@ internal void ResetRunspaceState(ExecutionContext context)
         }
 
-        internal static void SetSessionStateDrive(ExecutionContext context, bool setLocation)
+        private static bool TryInitSessionStateCurrentDrive(ExecutionContext context, string location)
```
`location` seems to only be used in the error message here, seems wrong to provide this value if the code isn't actually using it outside of an error. I don't think an error should be reported in a `Try*` method as well, the caller should be in charge of dealing with the error condition with this type of pattern.

jborean93 reviewed (commented):
The codefactor tests look like they need fixing, the PowerShell team won't really consider PRs ready until CI is green.

There should also be tests added for this new API to verify the new changes added. We should be testing things like

+ We can set it to a custom location
+ UNC paths work (Windows only and can use `\\localhost\...`)
+ Can set to a root of another drive, i.e. `Temp` and/or `Env`
+ Fails on invalid path (missing and invalid drive)

MatejKafka commented:
> The codefactor tests look like they need fixing, the PowerShell team won't really consider PRs ready until CI is green.

All of the codefactor issues are doc only, and most of the surrounding code is also breaking the rules. If you want me to fix the tests, I can, but I don't see any benefit from it. :)

> There should also be tests added for this new API to verify the new changes added. We should be testing things like

If the WG wants me to add unit tests, I can, but none of the other properties on `InitialSessionState` are currently unit tested.

jborean93 commented:
> All of the codefactor issues are doc only, and most of the surrounding code is also breaking the rules. If you want me to fix the tests, I can, but I don't see any benefit from it. :)

Unfortunately it's part of the CI tests, it needs to be fixed for it to be considered ready. It's part of the style tests of PowerShell and the documentation even derives what will be part of the docs when the it is live.

> If the WG wants me to add unit tests, I can, but none of the other properties on InitialSessionState are currently unit tested.

That doesn't mean you shouldn't add tests. Tests are critical to show that the code you've added works and also ensures there are no regressions in the future. I'm not part of the WG but you should always add tests to anything you touch regardless of whether or not there are existing ones.

MatejKafka committed:
InitialSessionState: Add unit tests for `Location`

MatejKafka commented:
I removed the exception docstrings, which fixes most of the codefactor issues. One of the remaining issues does not make any sense (spacing around `new(...)`), fixing the other one would make the docstring less readable and no other property in `InitialSessionState` uses the suggested pattern.

I also added a few unit tests for the `Location` property.

---

> Tests are critical to show that the code you've added works and also ensures there are no regressions in the future.

Maybe if PowerShell maintainers heeded that advice, I wouldn't be finding PowerShell bugs every other week, but I digress.

jborean93 commented:
> One of the remaining issues does not make any sense (spacing around new(...)

The rules that PowerShell uses don't like the `new()` syntax without the type name. You need to use `new TypeName(...);` instead.

> Maybe if PowerShell maintainers heeded that advice, I wouldn't be finding PowerShell bugs every other week, but I digress.

That's why adding tests are critical, we can't do much about existing tests but we can be sure that new features and fixes being added now have tests.

MatejKafka commented:
I don't believe the Pester test failures are related to my PR.

MatejKafka commented:
Bump. After rebase over current master, the CI is green.

MatejKafka commented:
@SeeminglyScience Could you please find some time to review this PR? It's been waiting for a year and I'd really like to finish it.
